### PR TITLE
Fix DMC batched driver dynamic population control

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Initial version to the function correct. Use algorithm mostly the same as the production code.
Add DMC/WalkerControl.cpp, a single implementation for MPI/non-MPI. Fix population scheme will be added in the same file.
The code path difference should be small.

Known issues:
1. Node walker number floor and ceiling have not been supported. MaxCopy is also not supported.
2. consecutive DMC sections stops by a a safeguard not allowing recreate walker controller. Should be an easy fix.
3. certain call flow optimization and refactoring are needed.
```
yeluo@epyc-server:~/opt/qmcpack/build_gcc_real$ ctest -R dmcbatch -j32 
Test project /home/yeluo/opt/qmcpack/build_gcc_real
    Start 1103: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_sdj-1-16
    Start 1105: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_4r_sdj-4-4
1/4 Test #1103: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_sdj-1-16 ...............   Passed   70.23 sec
    Start 1104: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_sdj-1-16-totenergy
2/4 Test #1104: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_sdj-1-16-totenergy .....   Passed    0.03 sec
3/4 Test #1105: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_4r_sdj-4-4 .............   Passed   71.83 sec
    Start 1106: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_4r_sdj-4-4-totenergy
4/4 Test #1106: short-diamondC_2x1x1_pp-vmcbatch_dmcbatch_4r_sdj-4-4-totenergy ...   Passed    0.03 sec
```

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
Epyc-server, Bora

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

**Comment by ye-luo** [comment]
@prckent I agree. Any floor or ceiling artificially applied on the population actually hurts the credibility of the statistics. If it happens, users should stay alerted and try to troubleshooting their TWF actually. What we should improve is alerting user if total walker count is hitting too low or too high and also alert users about too high branching multiplicity which may occur also in fixed population.

**Review by PDoakORNL — COMMENTED** [review]
I'm still working on my full review, looks good.

But WalkerControlMPI with respect to the unified driver was validated with test_WalkerControlMPI.cpp. These obviously never helped find whatever bug was killing the statistics when MPI balancing was occurring but did specify and validate the expected behavior. 

It seems like WalkerControl should satisfy these or similar unit tests.

This might answer the question of whether some of the behavior or WalkerControlMPI was actually invalid or what context or state was not included in the unit test and made bad code in WalkerControlMPI seem correct.

**Review by PDoakORNL — APPROVED** [review]
The clean transplant the "original" swapWalkersSimple might be what fixed this. But overall the code is shorter and cleaner now. The collapse an unnecessary inheritance pair and simplification code around measuring properties of the walkers makes especially seems to matter. While bringing what was worthwhile from my work on the walker control forward, thanks @ye-luo.  Unit tests really need to be returned for the methods the unified driver depends on, I don't think that should stop this PR.

File: src/QMCDrivers/DMC/DMCBatched.cpp [hunk 1]
```diff
@@ -550,7 +527,7 @@ bool DMCBatched::run()
           crowd_ref.accumulate(population_.get_num_global_walkers());
       }
 
-      branch_engine_->branch(step, population_);
+      branch_engine_->branch(block * qmcdriver_input_.get_max_steps() + step, population_);
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCBatched.cpp:530]: This only affected printing the index in dmc.dat.

**Comment by ye-luo** [comment]
@PDoakORNL Regarding unit tests. They will be added as I keep refactoring WalkerControl class. I still want to move a few variables from class scope to function scope for facilitating unit tests.

**Comment by prckent** [comment]
ClangDev complains:
```
warning: delete called on non-final 'qmcplusplus::WalkerControl' that has virtual functions but non-virtual destructor
```

**Pull request merged**